### PR TITLE
Implementing JWE for P256

### PR DESCRIPTION
**Pull request by bshambaugh**
I am starting to work through #225 and the first file that I changed was ECDH.ts. Please forgive my deletion of yarn.lock. If this is an issue, perhaps I can address it later. This does not have tests yet, but neither does the prexisting x25519 function. I do need to make sure that when I get a compressed P256 public key as input to getSharedSecret I get the correct results.

### CODE REVIEW

**Review by mirceanis — CHANGES_REQUESTED** [review]
Looks ok, thank you for adding this.
You need to add the lockfile back before the PR can be accepted.

**Comment by bshambaugh** [comment]
I did git `checkout upstream/master -- yarn.lock` ,  `git add yarn.lock ` , `git commit -m "add removed yarn.lock file"` , and `git push -u origin master ` to get to the present state. I thought that diff for yarn.lock would show now changes, but it still does.

**Comment by mirceanis** [comment]
> I did `git checkout upstream/master -- yarn.lock` , `git add yarn.lock ` , `git commit -m "add removed yarn.lock file"` , and `git push -u origin master ` to get to the present state on my fork. I thought that diff for yarn.lock would show no changes, but it still does.

Perhaps try to `git fetch --all` before the checkout

**Comment by bshambaugh** [comment]
I cloned this repository in a new directory and compared it to the fork and did what you said with git fetch and then did a local diff on yarn.lock. I even browsed through the file on github. The file yarn.lock doesn't appear to be different, yet it is showing differences in files changed for this PR.

**Comment by bshambaugh** [comment]
But, the last change in git log that is not by me is on sept 8th. On this upstream remote, the last change is sept 27th,

**Comment by bshambaugh** [comment]
`git merge origin/master upstream/master` then `git push` did the trick.

**Review by mirceanis — COMMENTED** [review]
I was trying to work with the p256 shared secret implementation from @noble/curves and discovered that it may have some differences when compared to other implementations (see https://github.com/paulmillr/noble-curves/issues/96)

My instinct here would be to align with the other implementations which would mean skipping the first byte of the output, but ideally this should all be based on some test vectors that ensure interoperability.

BTW, x25519 ECDH doesn't manifest the same problem as the key representation there is less flexible

**Comment by bshambaugh** [comment]
Awesome. I was gone for 2 weeks in the 1st half of October for and IIW et al. trip, but I have started to look at JWE again. This means I have been combing through the code base and searching the web and making notes in the past few weeks. 

tl;dr: good timing. more code may be coming soon.

**Comment by bshambaugh** [comment]
As I was looking through some repositories on didcomm I realized that there was some replication:
https://github.com/decentralized-identity/veramo/tree/next/packages/did-comm/src/encryption

This is not bad because I realized I was doing similar things
Screenshot_2023-11-22_14-04-20

Screenshot_2023-11-22_14-04-34

**Comment by mirceanis** [comment]
> As I was looking through some repositories on didcomm I realized that there was some replication

The code duplication from the did-comm implementation you linked to is not from my proudest moments..

> This is not bad because I realized I was doing similar things

There are definitely improvements to be made, so if you do manage to avoid duplication, that would be even better.


---


also maybe related: https://github.com/paulmillr/noble-ciphers/discussions/18

**Comment by bshambaugh** [comment]
Status update:

I'm trying to refactor code, but I am getting stuck. I apologize for the screenshots, but it seems to allow me to make this concise.

In
https://codesandbox.io/p/github/bshambaugh/did-jwt/master?file=%2Fsrc%2Fencryption%2[…]1%2C1-84%2C1&workspaceId=2a49a562-e345-4cca-bc1f-efecc0ea40b6 

Start also looks like the form in createEncrypterStart.ts

Finish looks like the form in createEncrytperFinal.ts 

start
finsh

createencrypterFinal.ts will be renamed to createEncrypter.ts

**Comment by bshambaugh** [comment]
There are multiple feature branches going on. Number [1]  just keeps the same **alg** and **enc** parameters as fox x25519 but changes the curve to P-256. 
Number [2]  tries to introduce ECDH-ES-A256KW as the **alg** and A256GCM as the **enc**.

For number [1] this still needs to be fixed:
https://github.com/bshambaugh/did-jwt/blob/X25519-ECDH-1PU-ES/src/encryption/xc20pEncryption.ts#L145-L191

For number [2]  a lot still needs to be done. You can see the present @noble and @stablelib having to be there to provide GCM and KW. https://github.com/bshambaugh/did-jwt/blob/as256kwExperiments/src/encryption/a256kwgcm.ts . 
https://github.com/bshambaugh/did-jwt/blob/as256kwExperiments/src/encryption/createEncrypter.ts will need to look like https://github.com/bshambaugh/did-jwt/blob/X25519-ECDH-1PU-ES/src/encryption/createEncrypter.ts ... 

& tests for both.

**Comment by bshambaugh** [comment]
https://github.com/bshambaugh/did-jwt/blob/X25519-ECDH-1PU-ES/src/encryption/createEncrypter.ts#L28 was added

`ephemeralKeyPair?: EphemeralKeyPair`

 to get
https://github.com/bshambaugh/did-jwt/blob/X25519-ECDH-1PU-ES/src/encryption/createEncrypter.ts#L80-L84

` if(ephemeralKeyPair?.publicKeyJWK.crv != null) { //uses added ephemeralKeyPair, the one in encryptCek not in scope
    return { alg: keyWrapper.alg, enc: contentEncrypter.enc, encrypt, encryptCek, genEpk: prefixToDriverMap[ephemeralKeyPair!.publicKeyJWK.crv!] } 
  } else {
    return { alg: keyWrapper.alg, enc: contentEncrypter.enc, encrypt, encryptCek, genEpk: genX25519EphemeralKeyPair }
  }`

to work. Except it doesn't yet. I need a better way to construct a conditional which depends on the cryptographic curve.

I added resolveP256Encrypters which is like resolveX25519Encrypters 

https://github.com/bshambaugh/did-jwt/blob/X25519-ECDH-1PU-ES/src/encryption/xc20pEncryption.ts#L449-L504
https://github.com/bshambaugh/did-jwt/blob/X25519-ECDH-1PU-ES/src/encryption/xc20pEncryption.ts#L145-L191

which is what @oed exported. https://github.com/oed/did-jwt/tree/fix/export-x25519-resolver .

**Comment by bshambaugh** [comment]
I fixed some stuff and All tests on my X25519-ECDH-1PU-ES branch appear to pass now.

File: src/__tests__/a256gcmEncryption.test.ts [hunk 1]
```diff
@@ -925,7 +926,197 @@ describe('ECDH-ES+A256KW (P-256) Auth, Key Wrapping Mode with A256GCM content en
...
+    })
+  })
+})
+
+// Adapted from:
+// https://github.com/decentralized-identity/veramo/blob/next/packages/did-comm/src/__tests__/encryption.test.ts#L492-L674
+// changes X25519 to P256
+describe('ECDH-ES+A256KW (P-256) Dir Auth, Key Wrapping Mode with A256GCM content encryption', () => {
+  describe('One recipient', () => {
+    let cleartext: Uint8Array, recipientKey: any, senderKey: any, decrypter: Decrypter, sharedSecret: Uint8Array
```
Comment (bshambaugh) [inline src/__tests__/a256gcmEncryption.test.ts:945]: Comment after DIF Hacakthon: It seems like unauthenticated encryption is the default for JWE with ECDH-ES+A256KW, so I should throw out any code that is '*'.Auth.'*' and just keep ''*'.Anon.'*'? 

i.e.: Keep AnonEncrypter, AnonDecrypter, AnonDirEncrypter and AnonDirDecrypter..

"JSON Object Signing and Encryption (JOSE) defines a number of
   encryption (JWE) [[RFC7516](https://datatracker.ietf.org/doc/html/rfc7516)] and digital signature (JWS) [[RFC7515](https://datatracker.ietf.org/doc/html/rfc7515)]
   algorithms.  When symmetric cryptography is used, JWE provides
   authenticated encryption that ensures both confidentiality and sender
   authentication.  However, for public key cryptography the existing
   JWE encryption algorithms provide only confidentiality and some level
   of ciphertext integrity.  When sender authentication is required,
   users must resort to nested signed-then-encrypted structures, which
   increases the overhead and size of resulting messages.  This document
   describes an alternative encryption algorithm called ECDH-1PU that
   provides public key authenticated encryption, allowing the benefits
   of authenticated encryption to be enjoyed for public key JWE as it
   currently is for symmetric cryptography."
   https://datatracker.ietf.org/doc/html/draft-madden-jose-ecdh-1pu-03

Comment (mirceanis) [reply src/__tests__/a256gcmEncryption.test.ts:945]: Yes, only `-1PU` is auth, `-ES` is anon

Comment (bshambaugh) [reply src/__tests__/a256gcmEncryption.test.ts]: to-do: make sure https://github.com/decentralized-identity/did-jwt/blob/master/src/__tests__/jwe-vectors.ts is up to date to cover these changes.

**Comment by bshambaugh** [comment]
not stale. still working on.
